### PR TITLE
[AXON-37] (PR-3) Implement headless E2E testing

### PR DESCRIPTION
**Pull request by sdzh-atlassian**
### What is this?

A much needed break from having to stare at the VSCode UI - at least for the duration of our end-to-end tests running 😉

In all seriousness, this is an implementation of dockerized local E2E test execution. This change features 3 new commands, the docker equivalent of previously implemented ones:

```
# Build the docker image for E2E testing.
# This only needs to be done once
npm run e2e:docker:build

# Run full cycle of test setup, including rebuilding the extension
# One would typically want to do this once, and then use the last command
npm run e2e:docker

# Skip most of the test setup, and run re-compiled tests
# MUCH faster, use this to iterate over tests themselves
npm run e2e:docker:rerun
```

The way this works is by building a (relatively) small docker image with `xvfb`, `vscode` and their deps - and mounting the whole working dir in there. The tests are then run using the virtual buffer provided by `xvfb-run`. Any failed tests will produce screenshots in the local folder `e2e/.resources/screenshots`. How it looks (there are no scary VSCode popups in the process 😉):

| When tests fail | When tests succeed |
|-|-|
|image | image |

### How was this tested

* All the new commands work
* I've built the `vsix` to make sure none of the extra stuff ends up there
* No changes to the code of the extension itself here - so didn't do any tests against that

### CODE REVIEW

File: e2e/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+# However, using `xvfb-run` as an entrypoint directly causes docker to hang (?!)
+# So, let's create an entrypoint to pass parameters to xvfb-run.
+# We'll do this in-place to not have any build context for this Dockerfile:
+RUN cat <<EOF > /usr/bin/entrypoint.sh && chmod +x /usr/bin/entrypoint.sh
+#!/usr/bin/env bash
+echo "Invoking xvfb-run with the following arguments: \$@"
+xvfb-run \$@
+EOF
+
+ENTRYPOINT ["/usr/bin/entrypoint.sh"]
```
Comment (sdzh-atlassian) [inline e2e/Dockerfile:41]: It took me. SO. LONG. to figure out that the issue I was having only had to do with improperly set up entrypoint 🙀 

This little built-in entrypoint script is the most elegant solution I could come up with to allow the flexibility of `docker run my-cool-app --with=fancy-options` without completely borking the whole process by using `xvfb-run` directly 😦

**Review by bwieger-atlassian-com — CHANGES_REQUESTED** [review]

File: package.json [hunk 2]
```diff
@@ -62,9 +62,12 @@
         "extension:package": "npm run extension:clean && vsce package --baseContentUrl https://raw.githubusercontent.com/atlassian/atlascode/main/ --allow-star-activation",
         "extension:install": "npm run extension:package && code --install-extension ./atlascode-*.vsix --force",
         "test": "jest",
+        "e2e": "npm run e2e:compile && e2e/scripts/run",
         "e2e:compile": "tsc --project e2e/tsconfig.e2e.json",
-        "e2e": "npm run e2e:compile && e2e/scripts/run.sh",
-        "e2e:rerun": "npm run e2e:compile && e2e/scripts/run.sh --rerun",
+        "e2e:rerun": "npm run e2e:compile && e2e/scripts/run --rerun",
+        "e2e:docker": "e2e/scripts/in-docker npm run e2e",
```
[the remark this answers is not in the document]
Comment (sdzh-atlassian) [reply package.json:68]: Addressed the command naming in the [follow-up PR](https://github.com/atlassian/atlascode/pull/36)

File: e2e/tests/dummy.test.ts [hunk 3]
```diff
@@ -1,28 +1,62 @@
...
+        ]);
+    });
+
+    it('should be installed', async () => {
+        const activityBar = new ActivityBar();
+        const controls = await activityBar.getViewControls();
+        // Get title from every control
+        const titles = await Promise.all(controls.map(async (control) => control.getTitle()));
+
+        expect('Atlassian').to.be.oneOf(titles);
```
[the remark this answers is not in the document]
Comment (sdzh-atlassian) [reply e2e/tests/dummy.test.ts:60]: On this specific line btw - I kind of screwed up here lol. It's supposed to be
```
expect(titles).to.include('Atlassian');
```

**Review by bwieger-atlassian-com — APPROVED** [review]

**Comment by sdzh-atlassian** [comment]
Force pushed to rebase to the previous branch in the chain, no changes, merging once the build is done
image

**Pull request merged**